src/metabase/api/search.clj

```diff
@@ -279,8 +279,16 @@
       (add-collection-join-and-where-clauses :collection.id search-ctx)))
 
 (s/defmethod search-query-for-model "database"
-  [model search-ctx :- SearchContext]
-  (base-query-for-model model search-ctx))
+  [model {:keys [current-user-perms] :as search-ctx} :- SearchContext]
```

Comment:
I think this could simply done by add another method for `check-permissions-for-model` with key is "database".

```clojure
(defmethod check-permissions-for-model :database
  [{:keys [id]}]
  (-> id Database mi/can-read?))
```
Your current approach works but it couples with the permissions logic.

Reply:
I had assumed since it was implemented for "table" in SQL here [here](https://github.com/metabase/metabase/blob/22695-fix-database-search-permissions/src/metabase/api/search.clj#L322) the permissions check was implemented in SQL for performance reasons. I haven't tested that assumption, but I will tomorrow as I would prefer `check-permissions-for-model` if we can afford it.

Reply:
Tested and it makes negligible difference to perf. I've implemented it with your suggestion.